### DISCUSSION TITLE
Correct way to detect whether the Console is open?

### DISCUSSION THREAD

**Discussion by luckman212** [Hammerspoon/hammerspoon#3079] [Q&A] [answered]
Anyone know the "correct" way to tell if the Hammerspoon Console is open?

The best I could come up with is this:

```lua
function isConsoleOpen()
  return hs.console.hswindow() ~= nil
end
```

**Comment by asmagill** [comment, marked as answer]
Any way that works is correct -- there is often more than one way to do something in Hammerspoon.

That said, I've found that some methods can sometimes seem to lag... I've found that the following seems to avoid the lag most (all? haven't seen it in a while) of the time:

~~~lua
function isConsoleOpen()
  local hspoon = hs.application.applicationsForBundleID(hs.processInfo.bundleID)[1]
  local conswin = hspoon:mainWindow()
  return conswin and true or false
end
~~~

FWIW, I use this in the following hot key which toggles to console for me -- if it's closed, open the console; if it's open but not frontmost, make it frontmost; if it's open and frontmost, close it -- it also tries to remember the window that was frontmost prior to bringing up the console so I go back to it when I toggle the console out.

~~~lua
local windowHolder
hs.hotkey.bind({"cmd", "alt", "ctrl"}, "r", function()
    local hspoon = hs.application.applicationsForBundleID(hs.processInfo.bundleID)[1]
    local conswin = hspoon:mainWindow()
    if conswin and hspoon:isFrontmost() then
        conswin:close()
        if windowHolder and #windowHolder:role() ~= 0 then
            windowHolder:becomeMain():focus()
        end
        windowHolder = nil
    else
        windowHolder = window.frontmostWindow()
        hs.openConsole()
    end
end, nil)
~~~

**Reply by luckman212** [reply]
@asmagill I'm a fan of `hs.application.applicationsForBundleID('foo')[1]` too. Thanks for sharing your function. That's exactly what I'm trying to do. Although in my case I'm binding it to `hs.dockIconClickCallback` instead of a hotkey.

**Reply by Rhys-T** [reply]
If you're trying to get to Hammerspoon in particular, `hs.application.applicationForPID(hs.processInfo.processID)` (or `hs.application(hs.processInfo.processID)` for short) will get the correct app directly, rather than looping through all running apps and checking their bundle IDs like [`applicationsForBundleID` does internally](https://github.com/Hammerspoon/hammerspoon/blob/56b5835dc1bf7bb430dfea3f1662379cccec0c82/Hammerspoon/HSuicore.m#L104-L115).

**Reply by unphased** [reply]
thank you. this is really a lot of complexity to get the hammerspoon console to open and close properly (closing is the hard part, i suspected we needed to cache the previous window to re-focus it... i was trying to look up 2nd-frontmost-window after closing the console, but that seems to be not possible. the frontmost app after closing the console is, of course, Hammerspoon).
